services/src/main/java/org/keycloak/protocol/ciba/decoupledauthn/DecoupledAuthenticationProviderBase.java

```diff
@@ -0,0 +1,262 @@
+        processor.setAuthenticationSession(authSession)
+                .setFlowId(flowId)
+                .setConnection(clientConnection)
+                .setEventBuilder(event)
+                .setRealm(realm)
+                .setSession(session)
+                .setUriInfo(session.getContext().getUri())
+                .setRequest(httpRequest);
+
+        processor.authenticateOnly();
+        processor.evaluateRequiredActionTriggers();
+        UserModel user = authSession.getAuthenticatedUser();
+        if (user.getRequiredActionsStream().count() > 0) {
+            event.error(Errors.RESOLVE_REQUIRED_ACTIONS);
+            throw new CorsErrorResponseException(cors, OAuthErrorException.INVALID_GRANT, "Account is not fully set up", Response.Status.BAD_REQUEST);
+        }
+
+        AuthenticationManager.setClientScopesInSession(authSession);
+
+        processor.attachSession();
+        UserSessionModel userSession = processor.getUserSession();
+        if (userSession == null) {
+            logger.info(" userSession is null.");
+        } else {
+            logger.info(" user session id = " + userSession.getId() + ", username = " + userSession.getUser().getUsername());
+        }
+        updateUserSessionFromClientAuth(userSession);
+
+        logger.info("  Created User Session's id                            = " + userSession.getId());
+        logger.info("  Submitted in advance User Session ID Will Be Created = " + getUserSessionIdWillBeCreated());
+
+        // authorization (consent)
+        UserConsentModel grantedConsent = session.users().getConsentByClient(realm, user.getId(), client.getId());
+        if (grantedConsent == null) {
+            grantedConsent = new UserConsentModel(client);
+            session.users().addConsent(realm, user.getId(), grantedConsent);
+            logger.info("  Consent updated : ");
+            grantedConsent.getGrantedClientScopes().stream().forEach(i->logger.info(i.getName()));
+        }
+
+        boolean updateConsentRequired = false;
+
+        for (String clientScopeId : authSession.getClientScopes()) {
+            ClientScopeModel clientScope = KeycloakModelUtils.findClientScopeById(realm, client, clientScopeId);
+            if (clientScope != null) {
+                if (!grantedConsent.isClientScopeGranted(clientScope) && clientScope.isDisplayOnConsentScreen()) {
+                    grantedConsent.addGrantedClientScope(clientScope);
+                    updateConsentRequired = true;
+                }
+            } else {
+                logger.warnf("Client scope or client with ID '%s' not found", clientScopeId);
+            }
+        }
+
+        if (updateConsentRequired) {
+            session.users().updateConsent(realm, user.getId(), grantedConsent);
+            logger.info("  Consent granted : ");
+            grantedConsent.getGrantedClientScopes().stream().forEach(i->logger.info(i.getName()));
```

Comment:
can be simplified to `grantedConsent.getGrantedClientScopes().forEach(i->logger.info(i.getName()));`

Reply:
@valb3r Yes, I will fix codes as you suggested.